File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
class Stack:
    def __init__(self):
        self.s = []
    def push(self, data):
        self.s.append(data)
    def pop(self):
        return self.s.pop()
    def is_empty(self):
        return len(self.s) == 0
    def peek(self):
        if self.is_empty() == True:
            return None
        else:
            return self.s[-1]
    def __str__(self):
        return str(self.s)

while True:
    S = input()
    if S == ".":
        break
    else:
        st = Stack()
        is_valid = True
        dot_cnt = 0
        for ch in S:
            if ch == "(":
                st.push(ch)
            elif ch == "[":
                st.push(ch)
            elif ch == ")":
                if st.peek() == "(":
                    st.pop()
                else:
                    is_valid = False
            elif ch == "]":
                if st.peek() == "[":
                    st.pop()
                else:
                    is_valid = False
            elif ch == ".":
                dot_cnt += 1
        
        if st.is_empty() and dot_cnt == 1 and S[-1] == "." and is_valid:
            print(f"yes")
        else:
            print(f"no")

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <iostream>
#include <string>
#include <stack>
using namespace std;
int main(void) {

	while (true) {
		string str;
		getline(cin, str);
		if (str == ".") break;

		stack<char> s;
		bool is_valid = true;

        for (auto c : str) {
            if (c == '(' || c == '[') {
                s.push(c);
            }
            else if (c == ')') {
                if (s.empty() || s.top() != '(') {
                    is_valid = false;
                    break;
                }
                s.pop();
            }
            else if (c == ']') {
                if (s.empty() || s.top() != '[') {
                    is_valid = false;
                    break;
                }
                s.pop();
            }
        }

		if (!s.empty()) is_valid = false;

		if (is_valid) cout << "yes\n";
		else cout << "no\n";
	}
}